src/Libraries/Microsoft.Extensions.AI/Functions/AIFunctionFactory.cs

```diff
@@ -181,35 +221,104 @@ public static ReflectionAIFunction Build(MethodInfo method, object? target, AIFu
+            }
+
+            if (method.DeclaringType is { } declaringType &&
+                !declaringType.IsAssignableFrom(targetType))
+            {
+                Throw.ArgumentException(nameof(targetType), "The target type must be assignable to the method's declaring type.");
+            }
+
+            return new(ReflectionAIFunctionDescriptor.GetOrCreate(method, options), targetType, options);
+        }
+
         private ReflectionAIFunction(ReflectionAIFunctionDescriptor functionDescriptor, object? target, AIFunctionFactoryOptions options)
         {
             FunctionDescriptor = functionDescriptor;
             Target = target;
             AdditionalProperties = options.AdditionalProperties ?? EmptyReadOnlyDictionary<string, object?>.Instance;
         }
 
+        private ReflectionAIFunction(
+            ReflectionAIFunctionDescriptor functionDescriptor,
+            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] Type targetType,
+            AIFunctionFactoryOptions options)
+        {
+            FunctionDescriptor = functionDescriptor;
+            TargetType = targetType;
+            AdditionalProperties = options.AdditionalProperties ?? EmptyReadOnlyDictionary<string, object?>.Instance;
+        }
+
         public ReflectionAIFunctionDescriptor FunctionDescriptor { get; }
         public object? Target { get; }
+        [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)]
+        public Type? TargetType { get; }
         public override IReadOnlyDictionary<string, object?> AdditionalProperties { get; }
         public override string Name => FunctionDescriptor.Name;
         public override string Description => FunctionDescriptor.Description;
         public override MethodInfo UnderlyingMethod => FunctionDescriptor.Method;
         public override JsonElement JsonSchema => FunctionDescriptor.JsonSchema;
         public override JsonSerializerOptions JsonSerializerOptions => FunctionDescriptor.JsonSerializerOptions;
 
-        protected override ValueTask<object?> InvokeCoreAsync(
+        protected override async ValueTask<object?> InvokeCoreAsync(
             AIFunctionArguments arguments,
             CancellationToken cancellationToken)
         {
-            var paramMarshallers = FunctionDescriptor.ParameterMarshallers;
-            object?[] args = paramMarshallers.Length != 0 ? new object?[paramMarshallers.Length] : [];
+            bool disposeTarget = false;
+            object? target = Target;
+            try
+            {
+                if (TargetType is { } targetType)
+                {
+                    Debug.Assert(target is null, "Expected target to be null when we have a non-null target type");
+                    Debug.Assert(!FunctionDescriptor.Method.IsStatic, "Expected an instance method");
 
-            for (int i = 0; i < args.Length; i++)
+                    target = arguments.Services is { } services ?
```

Comment:
I'm curious what this entails for lifetime management of `AIFunction` instances. If one of the constructor dependencies is transient, then the `AIFunction` itself must be transient but we won't be able to rely on DI to clean it up but instead the users would have to dispose it themselves.

Maybe an E2E example might better illuminate how this is supposed to be used, but my gut tells me that an approach where AIFunctions are directly inherited from using classes that are then registered offers better mechanical sympathy to existing DI mechanisms.

Reply:
>  If one of the constructor dependencies is transient, then the AIFunction itself must be transient

Why? This object is always new'd up fresh just for the invocation. It's effectively always transient.

That then has the same standard issue of not disposing of input from DI that might be singletons or the like, but that's nothing new.

According to @halter73, this is how MVC works as well.

Reply:
Oh ok, I had missed this was being done on each invocation.